File: drone_ws/src/drone_controllers/drone_controllers/mellinger_controller.py
#!usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped
from nav_msgs.msg import Odometry
from std_msgs.msg import Float64MultiArray
from numpy import cross, dot, linalg, array, eye, zeros, transpose
from math import cos, sin, sqrt
from scipy.spatial.transform import Rotation


class MellingerController(Node):
    def __init__(self):
        super().__init__("mellinger_controller")

        self.declare_parameter("mass", 0.5)
        self.declare_parameter("g", 9.81)
        self.declare_parameter("L",0.25)
        self.declare_parameter("kF", 3e-5)
        self.declare_parameter("kM", 1.1e-6)
        self.declare_parameter("Kv",10.0)
        self.declare_parameter("Kp",25.0)
        self.declare_parameter("Kr",1.0)
        self.declare_parameter("Kw",1.0)

        self.mass_ = self.get_parameter("mass").get_parameter_value().double_value
        self.g_ = self.get_parameter("g").get_parameter_value().double_value
        self.L_ = self.get_parameter("L").get_parameter_value().double_value
        self.kF_ = self.get_parameter("kF").get_parameter_value().double_value
        self.kM_ = self.get_parameter("kM").get_parameter_value().double_value

        self.get_logger().info("Using a mass of %.2f"%self.mass_)
        self.get_logger().info("Using a gravitational field strength of %.2f"%self.g_)
        self.get_logger().info("Using a Length difference between CM and Rotors of %.2f"%self.L_)
        self.get_logger().info("Using a Thrust Coefficient of %.5f"%self.kF_)
        self.get_logger().info("Using a Drag Coefficient of %.5f"%self.kM_)

        # Gets the time right now in nanoseconds
        self.Kv_ = eye(3) * self.get_parameter("Kv").get_parameter_value().integer_value
        # Controls the velocity correction, larger parameter means more correction
        self.Kp_ = eye(3) * self.get_parameter("Kp").get_parameter_value().integer_value
        # Controls the position correction, larger parameter means more correction
        self.Kr_ = eye(3) * self.get_parameter("Kr").get_parameter_value().integer_value
        # Controls the altitude correction, larger parameter means more correction
        self.Kw_ = eye(3) * self.get_parameter("Kw").get_parameter_value().integer_value
        # Controls the angular velocity correction, larger parameter means more correction

        self.motor_pub_ = self.create_publisher(Float64MultiArray,"motor_publisher/commands",10)
        # Sends the angular velocities to Gazebo [s1, s2, s3, s4]
        self.ground_truth_sub_ = self.create_subscription(Odometry, "mellinger_controller/odom",self.odomCallback, 10)
        # Receives s = {position, velocity, quaternion, angular velocity}
        self.velocity_sub_ = self.create_subscription(TwistStamped, "mellinger_controller/cmd_vel",self.velCallback, 10)
        # Sends the desired trajectory velocity and yaw

        self.M_ = array([
            [self.kF_,self.kF_,self.kF_,self.kF_],
            [0,-(self.kF_*self.L_),0,(self.kF_*self.L_)],
            [-(self.kF_*self.L_),0,(self.kF_*self.L_),0],
            [-self.kM_,self.kM_,-self.kM_,self.kM_]
        ])
        # Array matrix from (2)

        self.M_inv_ = linalg.inv(self.M_)
        # Inverse of matrix M

        self.r_ = zeros([3]) # Current position
        self.vel_ = zeros([3]) # Current velocity
        self.R_ = eye(3) # Current rotation matrix
        self.w_ = zeros([3]) # Current angular velocity

        self.vel_T_ = zeros([3]) # Desired velocity
        self.r_T_ = zeros([3]) # Desired position
        self.yaw_T_ = 0 # Desired yaw
        self.yaw_rate_T_ = 0.0# Desired yaw rate

        self.timer_ = self.create_timer(0.01, self.controlLoop) # Timer callbacks take no msg arguement


    def controlLoop(self): # Function that executes the full mellinger pipeline
        self.yaw_T_ += self.yaw_rate_T_ * 0.01
        # Desired yaw
        
        i_B = self.R_[:,0]
        h = array([i_B[0], i_B[1], 0])
        h = h / linalg.norm(h)
        h_T = array([cos(self.yaw_T_),sin(self.yaw_T_),0])

        Fdes = -self.Kp_ @ (self.r_ - self.r_T_) - self.Kv_ @ (self.vel_ - self.vel_T_) + (array([0,0,self.mass_*self.g_]))
        Pdes = dot(Fdes,self.R_[:,2])
        Rdes = zeros((3,3))
        Rdes[:,2] = Fdes / linalg.norm(Fdes)
        Rdes[:,0] = cross(cross(array([0,0,1]),h_T),Rdes[:,2]) / linalg.norm(cross(cross(array([0,0,1]),h_T),Rdes[:,2]))
        Rdes[:,1] = cross(Rdes[:,2],Rdes[:,0])

        temp_arr = (transpose(Rdes) @ self.R_ - transpose(self.R_) @ Rdes)
        eR = 0.5 * array([-temp_arr[1,0],temp_arr[0,2],-temp_arr[2,1]])

        h_w = zeros(3)
        W_T = zeros(3)
        W_T[0] =  0.0 # dot(-h_w,self.r_T_[:,1])
        W_T[1] = 0.0 # dot(h_w,self.r_T_[:,0])
        W_T[2] = (W_T[1] * dot(Rdes[:,2], cross(array([0,0,1]), h_T)) + self.yaw_rate_T_ * dot(h_T, Rdes[:,0])) / dot(cross(array([0,0,1]), h_T), Rdes[:,1])        
        eW = self.w_ - W_T
        Tdes = -self.Kr_ @ eR - self.Kw_ @ eW
        tau_body = transpose(self.R_) @ Tdes
        rotor_speed_sq = self.M_inv_ @ array([Pdes, tau_body[0],tau_body[1],tau_body[2]])
        rotor_speeds = array([sqrt(max(0,s)) for s in rotor_speed_sq])

        msg = Float64MultiArray()
        msg.data = rotor_speeds.tolist()
        self.motor_pub_.publish(msg)

    def odomCallback(self, msg:Odometry):
        self.r_ = array([msg.pose.pose.position.x, msg.pose.pose.position.y, msg.pose.pose.position.z])
        # Current xyz positions from the Odometry data in Gazebo or Rviz
        self.vel_ = array([msg.twist.twist.linear.x, msg.twist.twist.linear.y, msg.twist.twist.linear.z])
        # Current xyz velocities from the Odometry data in Gazebo or Rviz
        self.w_ = array([msg.twist.twist.angular.x, msg.twist.twist.angular.y, msg.twist.twist.angular.z])
        # Current xyz angular velocities from the Odometry data in Gazebo or Rviz
        self.R_ = Rotation.from_quat([msg.pose.pose.orientation.x, msg.pose.pose.orientation.y, msg.pose.pose.orientation.z,
                            msg.pose.pose.orientation.w]).as_matrix()
        # Current quaternion data from the Odometry data in Gazebo or Rviz

    def velCallback(self, msg:TwistStamped):
        self.vel_T_ = array([msg.twist.linear.x, msg.twist.linear.y, msg.twist.linear.z])
        # Array of the desired velocities in the xyz axis' 
        self.yaw_rate_T_ = msg.twist.angular.z
        # Desired yaw rate


def main(args=None):
    rclpy.init(args=args)
    # Initialising rclpy
    node = MellingerController()
    # Creating a node from the class we made
    rclpy.spin(node=node)
    # Spin the node so it stays active
    node.destroy_node()
    # Destroy the node once during keyboard interruption
    rclpy.shutdown()
    # Shutdown rclpy when no nodes are active

if __name__ == "__main__":
    main()
    # If running this file in a terminal whatever function is here will run

File: drone_ws/src/drone_controllers/src/mellinger_controller.cpp
#include "mellinger_controller.hpp"
#include <Eigen/Geometry>
using std::placeholders::_1;

MellingerController::MellingerController(const std::string& name): 
    Node(name)

{
    declare_parameter("mass",0.138);
    declare_parameter("g",9.81);
    declare_parameter("L",0.12);
    declare_parameter("kF",3.0e-5);
    declare_parameter("kM",1.1e-6);
    declare_parameter("Kp",2.0);
    declare_parameter("Kv",1.0);
    declare_parameter("Kr",0.1);
    declare_parameter("Kw",0.05);
    declare_parameter("show_logs",false);

    mass = get_parameter("mass").as_double();
    g = get_parameter("g").as_double();
    L = get_parameter("L").as_double();
    kF = get_parameter("kF").as_double();
    kM = get_parameter("kM").as_double();
    Kp = get_parameter("Kp").as_double() * Eigen::Matrix3d::Identity();
    Kv = get_parameter("Kv").as_double() * Eigen::Matrix3d::Identity();
    Kr = get_parameter("Kr").as_double() * Eigen::Matrix3d::Identity();
    Kw = get_parameter("Kw").as_double() * Eigen::Matrix3d::Identity();
    show_logs = get_parameter("show_logs").as_boolean();

    motor_pub_ = create_publisher<actuator_msgs::msg::Actuators>("/simple_velocity_controller/commands",10);
    ground_truth_sub_ = create_subscription<nav_msgs::msg::Odometry>("/mellinger_controller/odom",10,std::bind(&MellingerController::odomCallback,this,_1));
    velocity_sub_ = create_subscription<geometry_msgs::msg::TwistStamped>("/mellinger_controller/cmd_vel",10,std::bind(&MellingerController::velCallback,this,_1));
    
    broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(this);
    transform_stamped_.header.frame_id = "odom";
    transform_stamped_.child_frame_id = "root";
    int log_iterator = 0;

    M_ << kF << kF << kF << kF << 0 << -(kF*L) << 0 << (kF*L) << -(kF*L) << 0 << (kF*L) << 0 << -kM << kM << -kM << kM
    inv_M_ = M_.inverse();

    RCLCPP_INTO_STREAM(get_logger(), "C++ Implemenation of Mellinger Controller Running...")
    auto timer = rclcpp::create_timer(this, this->get_clock(), 0.01, std::bind(&MellingerController::controlLoop, this, _1));
}

void MellingerController::controlLoop(){
    log_iterator += 1;
    yaw_T_ += yaw_rate_T_ * 0.01;
    Eigen::Vector3d i_B = R_.col(0);
    Eigen::Vector3d h{i_B[0], i_B[1], 0};
    h /= h.norm();
    Eigen::RowVector3d h_T{std::cos(yaw_T_), std::sin(yaw_T_), 0.0};

    Eigen::RowVector3d weight{0.0, 0.0, mass * g};
    auto Fdes = -Kp * (r_ - r_T_) -Kv * (vel_ - vel_T_) + weight;
    auto Pdes = Fdes.dot(R_.col(2));
    Eigen::Matrix3d Rdes = Eigen::Matrix3d::Zero();
    Eigen::RowVector3d k{0.0, 0.0, 1.0};
    Rdes.col(2) = Fdes / Fdes.norm();
    Rdes.col(0) = Rdes.col(2).cross(k.cross(h_T));
    Rdes.col(1) = Rdes.col(2).cross(Rdes.col(0));

    Eigen::Matrix3d temp_arr = Rdes.transpose() * R_ - R_.transpose() * Rdes;
    Eigen::RowVector3d eR{temp_arr(1,0),temp_arr(0,2),temp_arr(2,1)};
    eR *= 0.5;
    Eigen::RowVector3d eW = w_;

    auto Tdes = -Kr * eR - Kw * eW;
    auto tau_body = R_.transpose() * Tdes;
    Eigen::Vector4d temp_arr_2{Pdes, tau_body.row(0), tau_body.row(1), tau_body.row(2)};
    auto rotor_speed_sq_ = inv_M_ * temp_arr_2;
    double rs[4]{std::max(0.0,rotor_speed_sq_.row(0)),std::max(0.0,rotor_speed_sq_.row(1)),std::max(0.0,rotor_speed_sq_.row(2),
    ),std::max(0.0,rotor_speed_sq_.row(3))};

    if(log_iterator % 20 == 0 && show_logs){
        RCLCPP_INTO_STREAM(get_logger(),"R_ = "<< R_);
        RCLCPP_INTO_STREAM(get_logger(),"eR = "<< eR);
        RCLCPP_INTO_STREAM(get_logger(),"eW = "<< eW);
        RCLCPP_INTO_STREAM(get_logger(),"Fdes = "<< Fdes);
        RCLCPP_INTO_STREAM(get_logger(),"Pdes = "<< Pdes);
        RCLCPP_INTO_STREAM(get_logger(),"Tdes = "<< Tdes);
        RCLCPP_INTO_STREAM(get_logger(),"tau_body = "<< tau_body);
        RCLCPP_INTO_STREAM(get_logger(),"rotor_speed_sq_ = "<< rotor_speed_sq_);
        RCLCPP_INTO_STREAM(get_logger(),"rotor_speeds = "<< rs);
    }

    actuator_msgs::msg::Actuators msg;
    msg.velocity(rs[0],rs[1],rs[2],rs[3]);
    motor_pub_->publish(msg);

}

void MellingerController::velCallback(const geometry_msgs::msg::TwistStamped& msg){
    r_T_ << msg.twist.linear.x << msg.twist.linear.y << msg.twist.linear.z
    yaw_T_ << msg.twist.angular.z
}

void MellingerController::odomCallback(const nav_msgs::msg::Odometry& msg){
    r_ << msg.pose.pose.position.x << msg.pose.pose.position.y << msg.pose.pose.position.z
    vel_ << msg.twist.twist.linear.x << msg.twist.twist.linear.y << msg.twist.twist.linear.z
    w_ << msg.twist.twist.angular.x << msg.twist.twist.angular.y << msg.twist.twist.angular.z
    Eigen::Quaterniond q(msg.pose.pose.orientation.x,msg.pose.pose.orientation.y,msg.pose.pose.orientation.z,
                        msg.pose.pose.orientation.w);
    R_ = q.toRotationMatrix();

    transform_stamped_.transform.translation.x = r_[0];
    transform_stamped_.transform.translation.y = r_[1];
    transform_stamped_.transform.translation.z = r_[2];
    transform_stamped_.transform.rotation.x = q.x();
    transform_stamped_.transform.rotation.y = q.y();
    transform_stamped_.transform.rotation.z = q.z();
    transform_stamped_.transform.rotation.w = q.w();
    broadcaster_->sendTransform(transform_stamped_);

}

int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MellingerController>("mellinger_controller");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}